ignition_ros2_control/src/ignition_system.cpp

```diff
@@ -0,0 +1,402 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "ignition_ros2_control/ignition_system.hpp"
+
+#include "ignition/gazebo/components/JointForce.hh"
+#include "ignition/gazebo/components/JointForceCmd.hh"
+#include "ignition/gazebo/components/JointPosition.hh"
+#include "ignition/gazebo/components/JointPositionReset.hh"
+#include "ignition/gazebo/components/JointVelocity.hh"
+#include "ignition/gazebo/components/JointVelocityCmd.hh"
+
+class ignition_ros2_control::IgnitionSystemPrivate
+{
+public:
+  IgnitionSystemPrivate() = default;
+
+  ~IgnitionSystemPrivate() = default;
+  /// \brief Degrees od freedom.
+  size_t n_dof_;
+
+  /// \brief last time the write method was called.
+  rclcpp::Time last_update_sim_time_ros_;
+
+  /// \brief vector with the joint's names.
+  std::vector<std::string> joint_names_;
+
+  /// \brief vector with the control method defined in the URDF for each joint.
+  std::vector<IgnitionSystemInterface::ControlMethod> joint_control_methods_;
+
+  /// \brief handles to the joints from within Gazebo
+  std::vector<ignition::gazebo::Entity> sim_joints_;
+
+  /// \brief vector with the current joint position
+  std::vector<double> joint_position_;
+
+  /// \brief vector with the current joint velocity
+  std::vector<double> joint_velocity_;
+
+  /// \brief vector with the current joint effort
+  std::vector<double> joint_effort_;
+
+  /// \brief vector with the current cmd joint position
+  std::vector<double> joint_position_cmd_;
+
+  /// \brief vector with the current cmd joint velocity
+  std::vector<double> joint_velocity_cmd_;
+
+  /// \brief vector with the current cmd joint effort
+  std::vector<double> joint_effort_cmd_;
+
+  /// \brief The current positions of the joints
+  std::vector<std::shared_ptr<hardware_interface::StateInterface>> joint_pos_state_;
+
+  /// \brief The current velocities of the joints
+  std::vector<std::shared_ptr<hardware_interface::StateInterface>> joint_vel_state_;
+
+  /// \brief The current effort forces applied to the joints
+  std::vector<std::shared_ptr<hardware_interface::StateInterface>> joint_eff_state_;
+
+  /// \brief The position command interfaces of the joints
+  std::vector<std::shared_ptr<hardware_interface::CommandInterface>> joint_pos_cmd_;
+
+  /// \brief The velocity command interfaces of the joints
+  std::vector<std::shared_ptr<hardware_interface::CommandInterface>> joint_vel_cmd_;
+
+  /// \brief The effort command interfaces of the joints
+  std::vector<std::shared_ptr<hardware_interface::CommandInterface>> joint_eff_cmd_;
+
+  ignition::gazebo::EntityComponentManager * ecm;
```

Comment:
Holding a pointer to the ECM is highly discouraged *. The ECM is passed to systems thorough the `*Update` calls and its constness changes according to the stage of simulation that we're in. For safe use of the ECM, systems should only use it within the update callbacks.

\* I know we have a couple of instances of this within `ign-gazebo` itself, but those are considered "advanced" usage in extreme cases

Reply:
I think I cannot change this. The reason why it's that we are using the `ros2_control` API with `controller_manager` to call read and write in the ignition plugin, there is no way no change the signature of these methods, but I'm only using the ecm in the `*Update calls` :angel:

Reply:
Gotcha. It may be worth it documenting it here that the ECM shouldn't be accessed outside those methods, otherwise things will go :boom: Someone looking at this code years from now may not know otherwise